src/main/java/org/intermine/biovalidator/parser/CsvParser.java

```diff
@@ -0,0 +1,107 @@
+package org.intermine.biovalidator.parser;
+
+/*
+ * Copyright (C) 2002-2019 FlyMine
+ *
+ * This code may be freely distributed and modified under the
+ * terms of the GNU Lesser General Public Licence.  This should
+ * be distributed with the code.  See the LICENSE file for more
+ * information or http://www.gnu.org/copyleft/lesser.html.
+ *
+ */
+
+import com.univocity.parsers.common.ParsingContext;
+import com.univocity.parsers.common.ResultIterator;
+import com.univocity.parsers.csv.CsvParserSettings;
+import org.apache.commons.lang3.StringUtils;
+import org.intermine.biovalidator.api.Parser;
+import org.intermine.biovalidator.api.ParsingException;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+/**
+ * Parser for parsing CSV/TSV files
+ * @author deepak kumar
+ */
+public class CsvParser implements Parser<String[]>
+{
+    private InputStreamReader inputStreamReader;
+    private boolean hasHeader;
+    private boolean allowComments;
+    private boolean shouldAutoDetectDelimiter;
+    private String delimiter;
+    private com.univocity.parsers.csv.CsvParser csvParser;
+    private ResultIterator<String[], ParsingContext> csvParserLineIterator;
+
+    /**
+     * Construct CsvParser with a input source and whether to allow comments or not
+     * and by auto-detecting delimiter
+     * @param inputStreamReader input source
+     * @param hasHeader whether first line is a header line or not
+     * @param allowComments whether to allow '#' based comments or not
+     */
+    public CsvParser(InputStreamReader inputStreamReader,
+                     boolean hasHeader,
+                     boolean allowComments) {
+        this(inputStreamReader, hasHeader, allowComments, "");
+    }
+
+    /**
+     * Construct CsvParser with a input source and a delimiter
+     * @param inputStreamReader input source
+     * @param hasHeader whether first line is a header line or not
+     * @param allowComments whether to allow '#' based comments or not
+     * @param delimiter delimiter to separate columns
+     */
+    public CsvParser(InputStreamReader inputStreamReader,
+                     boolean hasHeader,
+                     boolean allowComments,
+                     String delimiter) {
+        this.inputStreamReader = inputStreamReader;
+        this.hasHeader = hasHeader;
+        this.allowComments = allowComments;
+        this.delimiter = delimiter;
+
+        this.csvParser = createParser();
+        this.csvParserLineIterator = csvParser.iterate(inputStreamReader).iterator();
+    }
+
+    private com.univocity.parsers.csv.CsvParser createParser() {
+        CsvParserSettings settings = new CsvParserSettings();
+        if (hasHeader) {
```

Comment:
how accurate is this? what's the tradeoff between just skipping the first line always?

(not proposing that!! Just wondering! :))

Reply:
CsvHeaderDetector is an exact Java port of python std lib CSV module: here is the python docs : https://docs.python.org/3.1/library/csv.html#csv.Sniffer.has_header

Reply:
that's interesting.

So how accurate is it? What's it doing?

Reply:
I won't say it's super smart, what it does is simple:
1. It looks for a column which is of a single type(say integer) and if all rows have the same type of value except the first row, then the first row is considered as a header.

2. If the first check doesn't work, then it also counts the length of the value of each column, if all values of a column have the same length except the first row, then again its a header.

So, I ported this python lib's algorithm to Java just because I thought this should work fine in most cases and the second reason is its part of python standard lib. But it can easily fail if all values in CSV file are strings with different length.

Reply:
here's python documentation:
```pyton
 def has_header(self, sample):
        # Creates a dictionary of types of data in each column. If any
        # column is of a single type (say, integers), *except* for the first
        # row, then the first row is presumed to be labels. If the type
        # can't be determined, it is assumed to be a string in which case
        # the length of the string is the determining factor: if all of the
        # rows except for the first are the same length, it's a header.
        # Finally, a 'vote' is taken at the end for each column, adding or
        # subtracting from the likelihood of the first row being a header.
       ...
```